worker-thread.go

```diff
@@ -0,0 +1,255 @@
+	}
+}
+
+func tearDownWorkerScript(handler *workerThread, exitStatus int) {
+
+	// if the fake request is nil, no script was executed
+	if handler.fakeRequest == nil {
+		return
+	}
+
+	// if the worker request is not nil, the script might have crashed
+	// make sure to close the worker request context
+	if handler.workerRequest != nil {
+		fc := handler.workerRequest.Context().Value(contextKey).(*FrankenPHPContext)
+		maybeCloseContext(fc)
+		handler.workerRequest = nil
+	}
+
+	fc := handler.fakeRequest.Context().Value(contextKey).(*FrankenPHPContext)
+	fc.exitStatus = exitStatus
+
+	defer func() {
+		handler.fakeRequest = nil
+	}()
+
+	// on exit status 0 we just run the worker script again
+	worker := handler.worker
+	if fc.exitStatus == 0 {
+		// TODO: make the max restart configurable
+		metrics.StopWorker(worker.fileName, StopReasonRestart)
+		handler.backoff.recordSuccess()
+		if c := logger.Check(zapcore.DebugLevel, "restarting"); c != nil {
+			c.Write(zap.String("worker", worker.fileName))
+		}
+		return
+	}
+
+	// TODO: error status
+
+	// on exit status 1 we apply an exponential backoff when restarting
+	metrics.StopWorker(worker.fileName, StopReasonCrash)
+	if handler.backoff.recordFailure() {
+		if !watcherIsEnabled {
+			panic(fmt.Errorf("workers %q: too many consecutive failures", worker.fileName))
+		}
+		logger.Warn("many consecutive worker failures", zap.String("worker", worker.fileName), zap.Int("failures", handler.backoff.failureCount))
+	}
+}
+
+func (handler *workerThread) waitForWorkerRequest() bool {
+
+	if c := logger.Check(zapcore.DebugLevel, "waiting for request"); c != nil {
+		c.Write(zap.String("worker", handler.worker.fileName))
+	}
+
+	if handler.state.is(stateActive) {
+		metrics.ReadyWorker(handler.worker.fileName)
+		handler.state.set(stateReady)
+	}
+
+	var r *http.Request
+	select {
+	case <-handler.thread.drainChan:
+		if c := logger.Check(zapcore.DebugLevel, "shutting down"); c != nil {
+			c.Write(zap.String("worker", handler.worker.fileName))
+		}
+
+		// execute opcache_reset if the restart was triggered by the watcher
+		if watcherIsEnabled && handler.state.is(stateRestarting) {
+			C.frankenphp_reset_opcache()
+		}
+
+		return false
+	case r = <-handler.thread.requestChan:
+	case r = <-handler.worker.requestChan:
+	}
+
+	handler.workerRequest = r
+
+	if c := logger.Check(zapcore.DebugLevel, "request handling started"); c != nil {
+		c.Write(zap.String("worker", handler.worker.fileName), zap.String("url", r.RequestURI))
+	}
+
+	if err := updateServerContext(handler.thread, r, false, true); err != nil {
+		// Unexpected error
+		if c := logger.Check(zapcore.DebugLevel, "unexpected error"); c != nil {
+			c.Write(zap.String("worker", handler.worker.fileName), zap.String("url", r.RequestURI), zap.Error(err))
+		}
+		fc := r.Context().Value(contextKey).(*FrankenPHPContext)
+		rejectRequest(fc.responseWriter, err.Error())
+		maybeCloseContext(fc)
+		handler.workerRequest = nil
+		handler.thread.Unpin()
```

Comment:
Will this eventually cause a segfault on a long-running server? If I understand correctly, we have the actual worker script name pinned to this as well. `Unpin()` unpins **all** pinned memory, which would include the worker script name (and potentially anything else that was pinned while executing the worker script).

Since pinning seems to be attached to the request being performed, vs. the thread that is running, perhaps we should attach the pinned memory lifetime to the request vs. the thread?

Reply:
This is also the logic that's currently implemented, but I think that you could be right in some edge cases. The superglobals like `$_SERVER` from the 'fake' request get overwritten by an actual request and are not accessible afterwards. Pinning to the FrankenPHPContext might be the cleaner solution though.

Reply:
That's why I like this refactor; it made this logic more obvious.

Reply:
Actually, while thinking about it a bit more; I think it is fine to attach it to the thread, but then we should unpin as soon as we start waiting for a request and no earlier.

Reply:
Yeah I think you're right, otherwise this might be problematic if globals are accessed between 2 calls to `frankenphp_handle_request`. I'll change it to unpin when waiting, after script execution and on thread termination.